### PR TITLE
Make handling of debugger attributes more complete.

### PR DESCRIPTION
**Pull request by erozenfeld**
This change adds the following to handling of DebuggerDisplayAttribute and DebuggerTypeProxyAttribute:

1. The type proxy can be specified either as System.Type or as a string. This change handles the latter.
2. Both DebuggerDisplayAttribute and DebuggerTypeProxyAttribute can be specified at assembly level. In that case the attributes have either Target property (of type System.Type) or TargetTypeName property (of string type). This change handles these cases and marks the type proxies appropriately when the corresponding targets are marked.

### CODE REVIEW

**Comment by erozenfeld** [comment]
@marek-safar PTAL

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -610,8 +619,58 @@ protected virtual void DoAdditionalTypeProcessing (TypeDefinition method)
...
+		{
+			if (!assembly.HasCustomAttributes)
+				return;
+
+			foreach (CustomAttribute attribute in assembly.CustomAttributes) {
+				string attributeFullName = attribute.Constructor.DeclaringType.FullName;
+				switch (attributeFullName) {
+					case "System.Diagnostics.DebuggerDisplayAttribute":
+					case "System.Diagnostics.DebuggerTypeProxyAttribute":
+						if (_context.KeepMembersForDebuggerAttributes) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Extract this to special method so you don't need to do secondary attributeFullName check. Something like
`StoreDebuggerTypeTarget (_assemblyDebuggerDisplayAttributes, attribute);`

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 2]
```diff
@@ -610,8 +619,58 @@ protected virtual void DoAdditionalTypeProcessing (TypeDefinition method)
...
+							}
+						}
+						break;
+				}
+			}
+		}
+
+		void MarkTypeSpecialCustomAttributes(TypeDefinition type)
 		{
+			if (_assemblyDebuggerDisplayAttributes.ContainsKey (type)) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Use TryGetValue instead

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 3]
```diff
@@ -610,8 +619,58 @@ protected virtual void DoAdditionalTypeProcessing (TypeDefinition method)
...
+				MarkTypeWithDebuggerDisplayAttribute (type, _assemblyDebuggerDisplayAttributes[type]);
+			}
+
+			if (_assemblyDebuggerTypeProxyAttributes.ContainsKey (type)) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Use TryGetValue instead

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 4]
```diff
@@ -718,13 +777,26 @@ void MarkTypeWithDebuggerDisplayAttribute (TypeDefinition type, CustomAttribute
 		void MarkTypeWithDebuggerTypeProxyAttribute (TypeDefinition type, CustomAttribute attribute)
 		{
 			if (_context.KeepMembersForDebuggerAttributes) {
-				TypeReference proxyTypeReference = (TypeReference) attribute.ConstructorArguments [0].Value;
+				object constructorArgument = attribute.ConstructorArguments[0].Value;
+				TypeReference proxyTypeReference = null;
+				if (constructorArgument is TypeReference) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Use `as` instead

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 5]
```diff
@@ -718,13 +777,26 @@ void MarkTypeWithDebuggerDisplayAttribute (TypeDefinition type, CustomAttribute
...
+					proxyTypeReference = (TypeReference) constructorArgument;
+				}
+				else if (constructorArgument is string) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Use `as` instead

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 6]
```diff
@@ -610,8 +619,58 @@ protected virtual void DoAdditionalTypeProcessing (TypeDefinition method)
 		{
 		}
 
-		void MarkTypeSpecialCustomAttributes (TypeDefinition type)
+		void MarkAssemblySpecialCustomAttributes(AssemblyDefinition assembly)
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Follow coding style

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 7]
```diff
@@ -610,8 +619,58 @@ protected virtual void DoAdditionalTypeProcessing (TypeDefinition method)
...
+									}
+								}
+							}
+						}
+						break;
+				}
+			}
+		}
+
+		void MarkTypeSpecialCustomAttributes(TypeDefinition type)
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Follow coding style

**Comment by erozenfeld** [comment]
@marek-safar I addressed your comments. PTAL

**Pull request merged**